In [37]:
import pandas as pd 
import numpy as np 
import os
from pathlib import Path





### Data Paths

In [38]:
os.chdir(Path().resolve())

In [39]:
 # Use raw string (r"") for Windows paths
print(os.getcwd())  # Verify the change

C:\Users\felix\OneDrive\Desktop\Capstone\1\DSE6311OM_Group3\Notebook


In [40]:
grads = "../Data/gradrates.xlsx"
acctblty = "../Data/Accountability.xlsx"
percapita = "../Data/DOR_Income_EQV_Per_Capita.xlsx"
perpupil = "../Data/PerPupilExpenditures.xlsx" 
population = "../Data/selectedpopulations.xlsx" 
teacher = "../Data/teacherdata.xlsx" 
charterschools = "../Data/schooldistricts_charter.csv"
townschools="../Data/schooldistricts_Town.csv"

### Datasets
Convert Excel data sets to pandas dataframes to merge into one dataframe. - E.T.

In [41]:
df_grads = pd.read_excel(grads, header = 1)
df_acctblty = pd.read_excel(acctblty, header = 1)
df_percapita = pd.read_excel(percapita, header = 0)
df_perpupil = pd.read_excel(perpupil, header = 1)
df_population = pd.read_excel(population, header = 1)
df_teacher = pd.read_excel(teacher, header = 1)
df_charterschool = pd.read_csv(charterschools,header=0)
df_townschool = pd.read_csv(townschools,header=0)

In [42]:
df_grads

District Name  District Code  \
0         Abby Kelley Foster Charter Public (District)        4450000   
1                                             Abington          10000   
2    Academy Of the Pacific Rim Charter Public (Dis...        4120000   
3                                     Acton-Boxborough        6000000   
4    Advanced Math and Science Academy Charter (Dis...        4300000   
..                                                 ...            ...   
301                                         Winchester        3440000   
302                                           Winthrop        3460000   
303                                             Woburn        3470000   
304                                          Worcester        3480000   
305                                       State Totals              0   

    # in Cohort  % Graduated  % Still in School  % Non-Grad Completers  \
0            88        100.0                0.0                    0.0   
1           153         88.9                1.3                    0.7   
2            54         96.3                1.9                    0.0   
3           466         96.1                3.2                    0.2   
4           139         99.3                0.7                    0.0   
..          ...          ...                ...                    ...   
301         354         94.1                3.4                    0.3   
302         136         92.6                1.5                    0.0   
303         291         89.7                4.1                    0.3   
304       1,862         85.2                6.1                    1.1   
305      72,602         89.2                4.5                    0.9   

     % H.S. Equiv.  % Dropped Out  % Permanently Excluded  
0              0.0            0.0                     0.0  
1              0.0            9.2                     0.0  
2              0.0            1.9                     0.0  
3              0.0            0.4                     0.0  
4              0.0            0.0                     0.0  
..             ...            ...                     ...  
301            0.3            2.0                     0.0  
302            0.0            5.9                     0.0  
303            0.0            5.8                     0.0  
304            0.1            7.4                     0.0  
305            0.2            5.3                     0.0  

[306 rows x 9 columns]

In [43]:
df_acctblty

District Name  District Code  \
0         Abby Kelley Foster Charter Public (District)        4450000   
1                                             Abington          10000   
2    Academy Of the Pacific Rim Charter Public (Dis...        4120000   
3                                     Acton-Boxborough        6000000   
4                                             Acushnet          30000   
..                                                 ...            ...   
393                                           Winthrop        3460000   
394                                             Woburn        3470000   
395                                          Worcester        3480000   
396                                        Worthington        3490000   
397                                           Wrentham        3500000   

                       Overall Classification  \
0    Not requiring assistance or intervention   
1    Not requiring assistance or intervention   
2    Not requiring assistance or intervention   
3    Not requiring assistance or intervention   
4    Not requiring assistance or intervention   
..                                        ...   
393  Not requiring assistance or intervention   
394  Not requiring assistance or intervention   
395  Not requiring assistance or intervention   
396  Not requiring assistance or intervention   
397  Not requiring assistance or intervention   

                 Reason for Classification  \
0             Meeting or exceeding targets   
1         Moderate progress toward targets   
2         Moderate progress toward targets   
3             Meeting or exceeding targets   
4      Substantial progress toward targets   
..                                     ...   
393    Substantial progress toward targets   
394       Moderate progress toward targets   
395       Moderate progress toward targets   
396  Limited or no progress toward targets   
397    Substantial progress toward targets   

     Progress Toward Improvement Targets (%)  
0                                       82.0  
1                                       39.0  
2                                       48.0  
3                                       85.0  
4                                       58.0  
..                                       ...  
393                                     56.0  
394                                     32.0  
395                                     41.0  
396                                      9.0  
397                                     63.0  

[398 rows x 5 columns]

In [44]:
df_percapita

DOR Code  LEA Code Municipality      County  Cherry Sheet FY  Population  \
0           1         1     Abington    PLYMOUTH             2023       17062   
1           2         2        Acton   MIDDLESEX             2023       24021   
2           3         3     Acushnet     BRISTOL             2023       10559   
3           4         4        Adams   BERKSHIRE             2023        8166   
4           5         5       Agawam     HAMPDEN             2023       28692   
..        ...       ...          ...         ...              ...         ...   
346       347       347       Woburn   MIDDLESEX             2023       40876   
347       348       348    Worcester   WORCESTER             2023      206518   
348       349       349  Worthington   HAMPSHIRE             2023        1193   
349       350       350     Wrentham     NORFOLK             2023       12178   
350       351       351     Yarmouth  BARNSTABLE             2023       25023   

     DOR Income  DOR Income Per Capita          EQV  EQV Per Capita  
0     645765000                  37848   2379458500          139460  
1    1733168000                  72152   4940912200          205691  
2     376411000                  35648   1346443900          127516  
3     191139000                  23407    549332800           67271  
4     893064000                  31126   3286756800          114553  
..          ...                    ...          ...             ...  
346  1829233000                  44751   9226467200          225718  
347  4638107000                  22459  15426121200           74696  
348    34443000                  28871    182949300          153352  
349   738007000                  60602   2481220700          203746  
350   876692000                  35035   6880187700          274955  

[351 rows x 10 columns]

In [45]:
df_perpupil

District Name  District Code  \
0         Abby Kelley Foster Charter Public (District)        4450000   
1                                             Abington          10000   
2    Academy Of the Pacific Rim Charter Public (Dis...        4120000   
3                                     Acton-Boxborough        6000000   
4                                             Acushnet          30000   
..                                                 ...            ...   
393                                             Woburn        3470000   
394                                          Worcester        3480000   
395                                        Worthington        3490000   
396                                           Wrentham        3500000   
397                                       State Totals              0   

     In-District Expenditures Total In-district FTEs  \
0              $24,231,438.78                1,415.8   
1              $35,603,421.16                2,143.5   
2              $12,801,946.92                  467.4   
3             $100,564,755.40                5,181.7   
4              $14,679,700.00                  965.8   
..                        ...                    ...   
393            $94,093,579.73                4,289.4   
394           $501,140,676.40               24,319.6   
395             $1,904,264.00                   75.5   
396            $17,626,509.66                  920.9   
397        $18,401,964,825.00              865,749.1   

    In-District Expenditures per Pupil        Total Expenditures  \
0                           $17,114.65            $24,231,438.78   
1                           $16,609.95            $41,098,932.87   
2                           $27,392.63            $12,801,946.92   
3                           $19,407.68           $109,746,603.20   
4                           $15,199.52            $18,752,410.80   
..                                 ...                       ...   
393                         $21,936.30           $105,328,116.50   
394                         $20,606.45           $568,546,854.00   
395                         $25,222.04             $2,822,565.38   
396                         $19,140.53            $19,059,388.49   
397                         $21,255.54        $20,737,958,663.00   

       Total Pupil FTEs Total Expenditures per Pupil  
0               1,415.8                   $17,114.65  
1               2,297.3                   $17,890.10  
2                 467.4                   $27,392.63  
3               5,302.1                   $20,698.70  
4               1,254.7                   $14,945.73  
..                  ...                          ...  
393             4,516.7                   $23,319.71  
394            27,234.8                   $20,875.75  
395               142.1                   $19,863.23  
396               981.1                   $19,426.55  
397           947,578.4                   $21,885.22  

[398 rows x 8 columns]

In [46]:
df_population

District Name  District Code  \
0         Abby Kelley Foster Charter Public (District)        4450000   
1                                             Abington          10000   
2    Academy Of the Pacific Rim Charter Public (Dis...        4120000   
3                                     Acton-Boxborough        6000000   
4                                             Acushnet          30000   
..                                                 ...            ...   
394                                             Woburn        3470000   
395                                          Worcester        3480000   
396                                        Worthington        3490000   
397                                           Wrentham        3500000   
398                                       State Totals              0   

    High Needs#  High Needs% English Learners#  English Learners%  \
0         1,005         70.7               256               18.0   
1         1,029         47.3               236               11.0   
2           332         71.1                44                9.4   
3         1,538         29.6               317                6.2   
4           415         42.7                 6                0.6   
..          ...          ...               ...                ...   
394       2,332         53.6               461               10.8   
395      20,471         82.9             7,386               30.4   
396          41         53.9                 0                0.0   
397         284         29.8                24                2.5   
398     508,820         55.1           110,554               12.1   

    First Language Not English#  First Language Not English% Low Income#  \
0                           967                         68.0         837   
1                           362                         16.9         736   
2                           158                         33.8         273   
3                         1,141                         22.2         546   
4                            17                          1.8         320   
..                          ...                          ...         ...   
394                         894                         20.9       1,653   
395                      14,201                         58.4      17,636   
396                           0                          0.0          30   
397                          51                          5.4         135   
398                     228,169                         25.0     386,060   

     Low Income% Students with Disabilities# Students with Disabilities%  \
0           58.9                                                           
1           34.4                                                           
2           58.5                                                           
3           10.6                                                           
4           33.2                                                           
..           ...                         ...                         ...   
394         38.7                                                           
395         72.5                                                           
396         39.5                                                           
397         14.2                                                           
398         42.3                                                           

    Students with Disabilities#.1  Students with Disabilities%.1  
0                             166                           11.7  
1                             387                           17.8  
2                             115                           24.6  
3                             803                           15.5  
4                             154                           15.8  
..                            ...                            ...  
394                           896       

In [47]:
df_teacher

District Name  District Code  \
0         Abby Kelley Foster Charter Public (District)        4450000   
1                                             Abington          10000   
2    Academy Of the Pacific Rim Charter Public (Dis...        4120000   
3                                     Acton-Boxborough        6000000   
4                                             Acushnet          30000   
..                                                 ...            ...   
394                                             Woburn        3470000   
395                                          Worcester        3480000   
396                                        Worthington        3490000   
397                                           Wrentham        3500000   
398                                       State Totals              0   

    Total # of Teachers (FTE)  % of Teachers Licensed Student / Teacher Ratio  \
0                       119.6                    99.2               11.9 to 1   
1                       153.8                    97.1               13.9 to 1   
2                        49.6                    65.6                9.4 to 1   
3                       395.4                   100.0               13.0 to 1   
4                        76.0                   100.0               12.7 to 1   
..                        ...                     ...                     ...   
394                     374.6                    99.2               11.4 to 1   
395                   1,715.9                    97.6               14.2 to 1   
396                       9.2                   100.0                8.3 to 1   
397                      79.4                   100.0               12.0 to 1   
398                  76,977.6                    97.1               11.9 to 1   

     Percent of Experienced Teachers  \
0                               81.5   
1                               81.3   
2                               37.3   
3                               88.3   
4                               78.9   
..                               ...   
394                             85.7   
395                             77.3   
396                             69.6   
397                             85.3   
398                             81.3   

     Percent of Teachers without Waiver or Provisional License  \
0                                                 88.3           
1                                                 89.6           
2                                                 95.6           
3                                                 97.5           
4                                                 92.1           
..                                                 ...           
394                                               96.1           
395                                               89.8           
396                                               87.0           
397                                              100.0           
398                                               93.4           

     Percent Teaching In-Field  
0                         90.5  
1                         95.7  
2                         47.3  
3                         98.8  
4                         94.7  
..                         ...  
394                       95.7  
395                       93.0  
396                      100.0  
397                      100.0  
398                       92.3  

[399 rows x 8 columns]

In [48]:
df_population.isnull().sum() 

## the original dataset doesn't have the columns "Students with Disabilities# , Students with Disabilities% " 
## For this reason I checked why there is no data in these columns. It doesn't say null but it doesn't have any info either. 
## After I combined the datasets, I dropped these columns. - E.T.

District Name                    0
District Code                    0
High Needs#                      0
High Needs%                      0
English Learners#                0
English Learners%                0
First Language Not English#      0
First Language Not English%      0
Low Income#                      0
Low Income%                      0
Students with Disabilities#      0
Students with Disabilities%      0
Students with Disabilities#.1    0
Students with Disabilities%.1    0
dtype: int64

### Merge Datasets (LEFT join)

* I applied LEFT Join
* Keep the df_population datasets as left table since it has the highest number of rows(399)
* Then merged each datasets and dropped the "District Name" other wise it was creating duplicates - E.T.

In [49]:
# Start with base DataFrame
df = df_population.copy()

# Merge and drop redundant 'District Name' after each merge
df = df.merge(df_acctblty.drop(columns=['District Name']), on='District Code', how='left')
df = df.merge(df_perpupil.drop(columns=['District Name']), on='District Code', how='left')
df = df.merge(df_grads.drop(columns=['District Name']), on='District Code', how='left')
df = df.merge(df_teacher.drop(columns=['District Name']), on='District Code', how='left')

In [50]:
df.shape

(399, 36)

In [51]:
df.columns 

Index(['District Name', 'District Code', 'High Needs#', 'High Needs%',
       'English Learners#', 'English Learners%', 'First Language Not English#',
       'First Language Not English%', 'Low Income#', 'Low Income%',
       'Students with Disabilities#', 'Students with Disabilities%',
       'Students with Disabilities#.1', 'Students with Disabilities%.1',
       'Overall Classification', 'Reason for Classification',
       'Progress Toward Improvement Targets (%)', 'In-District Expenditures',
       'Total In-district FTEs', 'In-District Expenditures per Pupil',
       'Total Expenditures', 'Total Pupil FTEs',
       'Total Expenditures per Pupil', '# in Cohort', '% Graduated',
       '% Still in School', '% Non-Grad Completers', '% H.S. Equiv.',
       '% Dropped Out', '% Permanently Excluded', 'Total # of Teachers (FTE)',
       '% of Teachers Licensed', 'Student / Teacher Ratio',
       'Percent of Experienced Teachers',
       'Percent of Teachers without Waiver or Provisional L

In [52]:
df.isnull().sum()

District Name                                                 0
District Code                                                 0
High Needs#                                                   0
High Needs%                                                   0
English Learners#                                             0
English Learners%                                             0
First Language Not English#                                   0
First Language Not English%                                   0
Low Income#                                                   0
Low Income%                                                   0
Students with Disabilities#                                   0
Students with Disabilities%                                   0
Students with Disabilities#.1                                 0
Students with Disabilities%.1                                 0
Overall Classification                                        1
Reason for Classification               

In [53]:
df.dtypes

District Name                                                 object
District Code                                                  int64
High Needs#                                                   object
High Needs%                                                  float64
English Learners#                                             object
English Learners%                                            float64
First Language Not English#                                   object
First Language Not English%                                  float64
Low Income#                                                   object
Low Income%                                                  float64
Students with Disabilities#                                   object
Students with Disabilities%                                   object
Students with Disabilities#.1                                 object
Students with Disabilities%.1                                float64
Overall Classification            

### Combine DESE data with DOR data. - F.R.

In [54]:
df.head()
df_charterschool.head()
is_charter = df['District Name'].str.contains('charter', case=False, na=False)
df.loc[is_charter, 'CleanedName'] = df.loc[is_charter, 'District Name'].str.replace(r"\s*\(District\)", "", regex=True).str.strip().str.lower()

df_charterschool['CleanedName'] = (
    df_charterschool['NAME']
    .str.lower()
    .str.replace(r"school", "", regex=True)
    .str.replace(r"\s+", " ", regex=True)  
    .str.strip()
)


df = df.merge(
    df_charterschool[['CleanedName', 'MEMBERS']],
    how='left',
    left_on='CleanedName',
    right_on='CleanedName'
)


df_townschool['DISTRICT_N'] = df_townschool['DISTRICT_N'].str.strip()
#left join 
df = df.merge(
    df_townschool[['DISTRICT_N', 'MEMBERLIST']].rename(columns={'MEMBERLIST': 'MEMBERS_townschool'}),
    how='left',
    left_on='District Name',
    right_on='DISTRICT_N',
    suffixes=('', '_townschool')
)


# This step will populate if the school district has more than 1 towns
df['MEMBERS'] = df['MEMBERS'].fillna(df['MEMBERS_townschool'])
# This step is to populate when there is only member town and the member list is empty in schooltown ds
df['MEMBERS'] = df['MEMBERS'].fillna(df['District Name'])
df.drop(columns='DISTRICT_N', inplace=True)
df.drop(columns='MEMBERS_townschool', inplace=True)
df.head()

District Name  District Code  \
0       Abby Kelley Foster Charter Public (District)        4450000   
1                                           Abington          10000   
2  Academy Of the Pacific Rim Charter Public (Dis...        4120000   
3                                   Acton-Boxborough        6000000   
4                                           Acushnet          30000   

  High Needs#  High Needs% English Learners#  English Learners%  \
0       1,005         70.7               256               18.0   
1       1,029         47.3               236               11.0   
2         332         71.1                44                9.4   
3       1,538         29.6               317                6.2   
4         415         42.7                 6                0.6   

  First Language Not English#  First Language Not English% Low Income#  \
0                         967                         68.0         837   
1                         362                         16.9         736   
2                         158                         33.8         273   
3                       1,141                         22.2         546   
4                          17                          1.8         320   

   Low Income% Students with Disabilities# Students with Disabilities%  \
0         58.9                                                           
1         34.4                                                           
2         58.5                                                           
3         10.6                                                           
4         33.2                                                           

  Students with Disabilities#.1  Students with Disabilities%.1  \
0                           166                           11.7   
1                           387                           17.8   
2                           115                           24.6   
3                           803                           15.5   
4                           154                           15.8   

                     Overall Classification  \
0  Not requiring assistance or intervention   
1  Not requiring assistance or intervention   
2  Not requiring assistance or intervention   
3  Not requiring assistance or intervention   
4  Not requiring assistance or intervention   

             Reason for Classification  \
0         Meeting or exceeding targets   
1     Moderate progress toward targets   
2     Moderate progress toward targets   
3         Meeting or exceeding targets   
4  Substantial progress toward targets   

   Progress Toward Improvement Targets (%)  In-District Expenditures  \
0                                     82.0            $24,231,438.78   
1                                     39.0            $35,603,421.16   
2                                     48.0            $12,801,946.92   
3                                     85.0           $100,564,755.40   
4                                     58.0            $14,679,700.00   

  Total In-district FTEs In-District Expenditures per Pupil  \
0                1,415.8                         $17,114.65   
1                2,143.5                         $16,609.95   
2                  467.4                         $27,392.63   
3                5,181.7                         $19,407.68   
4                  965.8                         $15,199.52   

         Total Expenditures    Total Pupil FTEs Total Expenditures per Pupil  \
0            $24,231,438.78             1,415.8                   $17,114.65   
1            $41,098,932.87             2,297.3                   $17,890.10   
2            $12,801,946.92               467.4                   $27,392.63   
3           $109,746,603.20             5,302.1                   $20,698.70   
4            $18,752,410.80             1,254.7                   $14,945.73   

  # in Cohort  % Graduated  % Still in School  % Non-Grad Completers  \
0          88        

In [55]:
def get_avg_stats(members_str, df_percapita):
    
    towns = [t.strip() for t in members_str.split(',')] if pd.notnull(members_str) else []

    matched = df_percapita[df_percapita['Municipality'].isin(towns)]
 
    if not matched.empty:
        return pd.Series({
            'Population': matched['Population'].sum(),
            'DOR Income Per Capita': matched['DOR Income Per Capita'].mean()
        })
    else:
        return pd.Series({'Population': None, 'DOR Income Per Capita': None})


df[['Population', 'DOR Income Per Capita']] = df['MEMBERS'].apply(get_avg_stats, df_percapita=df_percapita)
df.head()

District Name  District Code  \
0       Abby Kelley Foster Charter Public (District)        4450000   
1                                           Abington          10000   
2  Academy Of the Pacific Rim Charter Public (Dis...        4120000   
3                                   Acton-Boxborough        6000000   
4                                           Acushnet          30000   

  High Needs#  High Needs% English Learners#  English Learners%  \
0       1,005         70.7               256               18.0   
1       1,029         47.3               236               11.0   
2         332         71.1                44                9.4   
3       1,538         29.6               317                6.2   
4         415         42.7                 6                0.6   

  First Language Not English#  First Language Not English% Low Income#  \
0                         967                         68.0         837   
1                         362                         16.9         736   
2                         158                         33.8         273   
3                       1,141                         22.2         546   
4                          17                          1.8         320   

   Low Income% Students with Disabilities# Students with Disabilities%  \
0         58.9                                                           
1         34.4                                                           
2         58.5                                                           
3         10.6                                                           
4         33.2                                                           

  Students with Disabilities#.1  Students with Disabilities%.1  \
0                           166                           11.7   
1                           387                           17.8   
2                           115                           24.6   
3                           803                           15.5   
4                           154                           15.8   

                     Overall Classification  \
0  Not requiring assistance or intervention   
1  Not requiring assistance or intervention   
2  Not requiring assistance or intervention   
3  Not requiring assistance or intervention   
4  Not requiring assistance or intervention   

             Reason for Classification  \
0         Meeting or exceeding targets   
1     Moderate progress toward targets   
2     Moderate progress toward targets   
3         Meeting or exceeding targets   
4  Substantial progress toward targets   

   Progress Toward Improvement Targets (%)  In-District Expenditures  \
0                                     82.0            $24,231,438.78   
1                                     39.0            $35,603,421.16   
2                                     48.0            $12,801,946.92   
3                                     85.0           $100,564,755.40   
4                                     58.0            $14,679,700.00   

  Total In-district FTEs In-District Expenditures per Pupil  \
0                1,415.8                         $17,114.65   
1                2,143.5                         $16,609.95   
2                  467.4                         $27,392.63   
3                5,181.7                         $19,407.68   
4                  965.8                         $15,199.52   

         Total Expenditures    Total Pupil FTEs Total Expenditures per Pupil  \
0            $24,231,438.78             1,415.8                   $17,114.65   
1            $41,098,932.87             2,297.3                   $17,890.10   
2            $12,801,946.92               467.4                   $27,392.63   
3           $109,746,603.20             5,302.1                   $20,698.70   
4            $18,752,410.80             1,254.7                   $14,945.73   

  # in Cohort  % Graduated  % Still in School  % Non-Grad Completers  \
0          88        

### Drop columsn that are not needed for analysis - these columns represent the count of students percentage of students will be used for these features as it is more useful for comparative analysis make these changes in df (inplace = True) - A.F.

In [56]:

df.drop([
    'High Needs#', 
    'English Learners#', 
    'First Language Not English#', 
    'Low Income#', 'Students with Disabilities#.1', 
    'CleanedName', 
    '# in Cohort', 
    'District Code'], 
    axis = 1, inplace=True)

# expand the display options for better visibility of the dataframe - A.F.
pd.set_option('display.max_columns', 100)
pd.set_option('display.max_rows', 300)

# check changes made to the dataframe - A.F.
df.head()
df.shape

(399, 32)

### Drop rows where graduation rate = NaN; these districts do not include high school grades and feed into other districts for grades 9-12 
### 93 rows will be dropped - A.F.

In [57]:
df.dropna(subset = ["% Graduated"], inplace = True)
df.shape

(306, 32)

### Check column names in the dataframe to see where changes are needed 
### Rename columns to be more readable and consistent with Python naming conventions- E.T.

In [58]:
df.columns

Index(['District Name', 'High Needs%', 'English Learners%',
       'First Language Not English%', 'Low Income%',
       'Students with Disabilities#', 'Students with Disabilities%',
       'Students with Disabilities%.1', 'Overall Classification',
       'Reason for Classification', 'Progress Toward Improvement Targets (%)',
       'In-District Expenditures', 'Total In-district FTEs',
       'In-District Expenditures per Pupil', 'Total Expenditures',
       'Total Pupil FTEs', 'Total Expenditures per Pupil', '% Graduated',
       '% Still in School', '% Non-Grad Completers', '% H.S. Equiv.',
       '% Dropped Out', '% Permanently Excluded', 'Total # of Teachers (FTE)',
       '% of Teachers Licensed', 'Student / Teacher Ratio',
       'Percent of Experienced Teachers',
       'Percent of Teachers without Waiver or Provisional License',
       'Percent Teaching In-Field', 'MEMBERS', 'Population',
       'DOR Income Per Capita'],
      dtype='object')

In [59]:
df.rename(columns={
    'District Name': 'district_name',
    'High Needs%' : 'high_needs_pct',
    'English Learners%' : 'english_learners_pct',
    'First Language Not English%' : 'first_language_not_english_pct',
    'Low Income%': 'low_income_pct',
    'Students with Disabilities%.1': 'students_with_disabilities_pct',
    'Overall Classification': 'overall_classification',
    'Reason for Classification': 'reason_for_classification',
    'Progress Toward Improvement Targets (%)': 'progress_toward_improvement_targets_pct',
    'In-District Expenditures': 'in_district_expenditures',
    'Total In-district FTEs': 'total_in_district_FTEs',
    'In-District Expenditures per Pupil': 'in_district_expenditures_per_pupil',
    'Total Expenditures': 'total_expenditures',
    'Total Pupil FTEs': 'total_pupil_FTEs',
    'Total Expenditures per Pupil': 'total_expenditures_per_pupil',
    '% Graduated': 'graduation_rate_pct',
    '% Still in School': 'still_in_school_pct',
    '% Non-Grad Completers': 'non_grad_completers_pct',
    '% H.S. Equiv.': 'hs_equivalency_pct',
    '% Dropped Out': 'dropout_rate_pct',
    '% Permanently Excluded': 'permanently_excluded_pct',
    'Total # of Teachers (FTE)': 'total_teachers_FTE',
    '% of Teachers Licensed': 'teachers_licensed_pct',
    'Student / Teacher Ratio': 'student_teacher_ratio',
    'Percent of Experienced Teachers': 'experienced_teachers_pct',
    'Percent of Teachers without Waiver or Provisional License': 'teachers_without_waiver_pct',
    'Percent Teaching In-Field': 'teaching_in_field_pct',
    'MEMBERS': 'members', 
    'Population': 'population', 
    'DOR Income Per Capita' : 'DOR_income_per_capita'}, inplace=True)

In [60]:
# check the changes made to the dataframe
df.head()

district_name  high_needs_pct  \
0       Abby Kelley Foster Charter Public (District)            70.7   
1                                           Abington            47.3   
2  Academy Of the Pacific Rim Charter Public (Dis...            71.1   
3                                   Acton-Boxborough            29.6   
5  Advanced Math and Science Academy Charter (Dis...            26.0   

   english_learners_pct  first_language_not_english_pct  low_income_pct  \
0                  18.0                            68.0            58.9   
1                  11.0                            16.9            34.4   
2                   9.4                            33.8            58.5   
3                   6.2                            22.2            10.6   
5                   1.4                            23.6            16.0   

  Students with Disabilities# Students with Disabilities%  \
0                                                           
1                                                           
2                                                           
3                                                           
5                                                           

   students_with_disabilities_pct                    overall_classification  \
0                            11.7  Not requiring assistance or intervention   
1                            17.8  Not requiring assistance or intervention   
2                            24.6  Not requiring assistance or intervention   
3                            15.5  Not requiring assistance or intervention   
5                             6.8  Not requiring assistance or intervention   

          reason_for_classification  progress_toward_improvement_targets_pct  \
0      Meeting or exceeding targets                                     82.0   
1  Moderate progress toward targets                                     39.0   
2  Moderate progress toward targets                                     48.0   
3      Meeting or exceeding targets                                     85.0   
5      Meeting or exceeding targets                                     82.0   

   in_district_expenditures total_in_district_FTEs  \
0            $24,231,438.78                1,415.8   
1            $35,603,421.16                2,143.5   
2            $12,801,946.92                  467.4   
3           $100,564,755.40                5,181.7   
5            $15,065,009.41                  964.3   

  in_district_expenditures_per_pupil        total_expenditures  \
0                         $17,114.65            $24,231,438.78   
1                         $16,609.95            $41,098,932.87   
2                         $27,392.63            $12,801,946.92   
3                         $19,407.68           $109,746,603.20   
5                         $15,623.23            $15,065,009.41   

     total_pupil_FTEs total_expenditures_per_pupil  graduation_rate_pct  \
0             1,415.8                   $17,114.65                100.0   
1             2,297.3                   $17,890.10                 88.9   
2               467.4                   $27,392.63                 96.3   
3             5,302.1                   $20,698.70                 96.1   
5               964.3                   $15,623.23                 99.3   

   still_in_school_pct  non_grad_completers_pct  hs_equivalency_pct  \
0                  0.0                      0.0                 0.0   
1                  1.3                      0.7                 0.0   
2                  1.9                      0.0                 0.0   
3                  3.2                      0.2                 0.0   
5                  0.7                      0.0                 0.0   

   dropout_rate_pct  permanently_excluded_pct total_teachers_FTE  \
0               0.0                       0.0              119.6   
1               9.2                       0.0              153.8   
2               1.9         

### Change the Student / teacher ratio values to the numeric by keeping the first decimal value as data. - E.T.

In [61]:
df["student_teacher_ratio"] = (
    df["student_teacher_ratio"]
    .astype(str)
    .str.extract(r'([\d\.]+)')  # Extract numeric part before 'to'
    .astype(float)
)

In [62]:
df.head()

district_name  high_needs_pct  \
0       Abby Kelley Foster Charter Public (District)            70.7   
1                                           Abington            47.3   
2  Academy Of the Pacific Rim Charter Public (Dis...            71.1   
3                                   Acton-Boxborough            29.6   
5  Advanced Math and Science Academy Charter (Dis...            26.0   

   english_learners_pct  first_language_not_english_pct  low_income_pct  \
0                  18.0                            68.0            58.9   
1                  11.0                            16.9            34.4   
2                   9.4                            33.8            58.5   
3                   6.2                            22.2            10.6   
5                   1.4                            23.6            16.0   

  Students with Disabilities# Students with Disabilities%  \
0                                                           
1                                                           
2                                                           
3                                                           
5                                                           

   students_with_disabilities_pct                    overall_classification  \
0                            11.7  Not requiring assistance or intervention   
1                            17.8  Not requiring assistance or intervention   
2                            24.6  Not requiring assistance or intervention   
3                            15.5  Not requiring assistance or intervention   
5                             6.8  Not requiring assistance or intervention   

          reason_for_classification  progress_toward_improvement_targets_pct  \
0      Meeting or exceeding targets                                     82.0   
1  Moderate progress toward targets                                     39.0   
2  Moderate progress toward targets                                     48.0   
3      Meeting or exceeding targets                                     85.0   
5      Meeting or exceeding targets                                     82.0   

   in_district_expenditures total_in_district_FTEs  \
0            $24,231,438.78                1,415.8   
1            $35,603,421.16                2,143.5   
2            $12,801,946.92                  467.4   
3           $100,564,755.40                5,181.7   
5            $15,065,009.41                  964.3   

  in_district_expenditures_per_pupil        total_expenditures  \
0                         $17,114.65            $24,231,438.78   
1                         $16,609.95            $41,098,932.87   
2                         $27,392.63            $12,801,946.92   
3                         $19,407.68           $109,746,603.20   
5                         $15,623.23            $15,065,009.41   

     total_pupil_FTEs total_expenditures_per_pupil  graduation_rate_pct  \
0             1,415.8                   $17,114.65                100.0   
1             2,297.3                   $17,890.10                 88.9   
2               467.4                   $27,392.63                 96.3   
3             5,302.1                   $20,698.70                 96.1   
5               964.3                   $15,623.23                 99.3   

   still_in_school_pct  non_grad_completers_pct  hs_equivalency_pct  \
0                  0.0                      0.0                 0.0   
1                  1.3                      0.7                 0.0   
2                  1.9                      0.0                 0.0   
3                  3.2                      0.2                 0.0   
5                  0.7                      0.0                 0.0   

   dropout_rate_pct  permanently_excluded_pct total_teachers_FTE  \
0               0.0                       0.0              119.6   
1               9.2                       0.0              153.8   
2               1.9         

### Fill in null values under 'population' column for regional school districts.  Data for these missing values are in the "Missing Regional Schools" table, which is merged with the main dataframe on a left join. - A.F.

In [63]:
df_updates = pd.read_excel("Missing Regional Schools.xlsx")
df_updates

FileNotFoundError: [Errno 2] No such file or directory: 'Missing Regional Schools.xlsx'

In [ ]:
df_updates = df_updates.rename(columns={"District Name": "district_name"})
df_merged = df.merge(df_updates, on="district_name", how="left", suffixes=("", "_update"))

In [ ]:
df_merged["population"] = df_merged["population"].fillna(df_merged["population_update"])
df_merged["DOR_income_per_capita"] = df_merged["DOR_income_per_capita"].fillna(df_merged["DOR_income_per_capita_update"])
df_merged["members"]=df_merged["members"].fillna(df_merged["members_update"])

In [ ]:
df_merged.head(20)

In [ ]:
df_final = df_merged.drop(columns=["population_update", "DOR_income_per_capita_update", "members_update"])
df_final = df_final.set_index("district_name")

In [ ]:
df=df_final

### Save the Combined Dataset - A.F.

In [ ]:
# save the final dataframe to a new Excel file
output_file = "combined_data.xlsx"
df.to_excel(output_file, index= True)

In [ ]:
# print the first few rows of the final dataframe
pd.set_option('display.max_columns', None)
df.head()